From cea203aefcd63400f4ab8239a6def11d592922de Mon Sep 17 00:00:00 2001
From: Sebastian Bazley <sebb@apache.org>
Date: Sun, 10 Apr 2011 11:26:55 +0000
Subject: [PATCH] IO-263 FileSystemUtils.freeSpaceKb throws exception for
 Windows volumes with no visible files. Improve coverage by also looking for
 hidden files.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/io/trunk@1090779 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                                   | 4 ++++
 src/main/java/org/apache/commons/io/FileSystemUtils.java  | 2 +-
 .../org/apache/commons/io/FileSystemUtilsTestCase.java    | 8 ++++----
 3 files changed, 9 insertions(+), 5 deletions(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index d99acf1d892..a12c1094bf9 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -40,6 +40,10 @@ The <action> type attribute can be add,update,fix,remove.
 
   <body>
     <release version="1.4" date="Not yet released">
+      <action dev="sebb" type="fix" issue="IO-263" due-to="Gil Adam">
+        FileSystemUtils.freeSpaceKb throws exception for Windows volumes with no visible files.
+        Improve coverage by also looking for hidden files.
+      </action>
     </release>
 
     <release version="1.3.2" date="2007-Jul-02">
diff --git a/src/main/java/org/apache/commons/io/FileSystemUtils.java b/src/main/java/org/apache/commons/io/FileSystemUtils.java
index 11cf20aae92..49eef97c779 100644
--- a/src/main/java/org/apache/commons/io/FileSystemUtils.java
+++ b/src/main/java/org/apache/commons/io/FileSystemUtils.java
@@ -297,7 +297,7 @@ long freeSpaceWindows(String path, long timeout) throws IOException {
         }
         
         // build and run the 'dir' command
-        String[] cmdAttribs = new String[] {"cmd.exe", "/C", "dir /-c " + path};
+        String[] cmdAttribs = new String[] {"cmd.exe", "/C", "dir /a /-c " + path};
         
         // read in the output of the command to an ArrayList
         List<String> lines = performCommand(cmdAttribs, Integer.MAX_VALUE, timeout);
diff --git a/src/test/java/org/apache/commons/io/FileSystemUtilsTestCase.java b/src/test/java/org/apache/commons/io/FileSystemUtilsTestCase.java
index be68a713836..ac489f482fe 100644
--- a/src/test/java/org/apache/commons/io/FileSystemUtilsTestCase.java
+++ b/src/test/java/org/apache/commons/io/FileSystemUtilsTestCase.java
@@ -186,7 +186,7 @@ public void testGetFreeSpaceWindows_String_EmptyPath() throws Exception {
             "17/08/2005  21:44    <DIR>          Desktop\n" +
             "               7 File(s)         180260 bytes\n" +
             "              10 Dir(s)     41411551232 bytes free";
-        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /-c ");
+        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /a /-c ");
         assertEquals(41411551232L, fsu.freeSpaceWindows("", -1));
     }
 
@@ -203,7 +203,7 @@ public void testGetFreeSpaceWindows_String_NormalResponse() throws Exception {
             "17/08/2005  21:44    <DIR>          Desktop\n" +
             "               7 File(s)         180260 bytes\n" +
             "              10 Dir(s)     41411551232 bytes free";
-        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /-c \"C:\"");
+        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /a /-c \"C:\"");
         assertEquals(41411551232L, fsu.freeSpaceWindows("C:", -1));
     }
 
@@ -220,7 +220,7 @@ public void testGetFreeSpaceWindows_String_StripDrive() throws Exception {
             "17/08/2005  21:44    <DIR>          Desktop\n" +
             "               7 File(s)         180260 bytes\n" +
             "              10 Dir(s)     41411551232 bytes free";
-        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /-c \"C:\\somedir\"");
+        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /a /-c \"C:\\somedir\"");
         assertEquals(41411551232L, fsu.freeSpaceWindows("C:\\somedir", -1));
     }
 
@@ -237,7 +237,7 @@ public void testGetFreeSpaceWindows_String_quoted() throws Exception {
             "17/08/2005  21:44    <DIR>          Desktop\n" +
             "               7 File(s)         180260 bytes\n" +
             "              10 Dir(s)     41411551232 bytes free";
-        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /-c \"C:\\somedir\"");
+        FileSystemUtils fsu = new MockFileSystemUtils(0, lines, "dir /a /-c \"C:\\somedir\"");
         assertEquals(41411551232L, fsu.freeSpaceWindows("\"C:\\somedir\"", -1));
     }
 
